core/src/main/java/com/linecorp/armeria/server/HttpHeaderPathMapping.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.StringJoiner;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.util.Exceptions;
+
+/**
+ * A {@link PathMapping} based on {@link HttpMethod} and {@link MediaType}.
+ */
+final class HttpHeaderPathMapping implements PathMapping {
+
+    private static final HttpResponseException METHOD_NOT_ALLOWED_EXCEPTION =
+            Exceptions.clearTrace(new HttpResponseException(HttpStatus.METHOD_NOT_ALLOWED));
+    private static final HttpResponseException NOT_ACCEPTABLE_EXCEPTION =
+            Exceptions.clearTrace(new HttpResponseException(HttpStatus.NOT_ACCEPTABLE));
+    private static final HttpResponseException UNSUPPORTED_MEDIA_TYPE_EXCEPTION =
+            Exceptions.clearTrace(new HttpResponseException(HttpStatus.UNSUPPORTED_MEDIA_TYPE));
```

Comment:
Please respect `Flags.verboseException()`:

```java
private HttpResponseException newMethodNotAllowedException() {
    if (Flags.verboseException()) {
        return new HttpResponseException(HttpStatus.METHOD_NOT_ALLOWED);
    } else {
        return METHOD_NOT_ALLOWED_EXCEPTION;
    }
}
```

Alternatively, we could generalize this in `HttpResponseException`:

```java
public class HttpResponseException ... {

    public static HttpResponseException of(HttpStatus httpStatus) {
        if (Flags.verboseException()) {
            return new HttpResponseException(httpStatus);
        } else {
            // do some lazy initialization using the cache backed by Int2ObjectMap.
        }
    }
    ...
}
```